### PR TITLE
fix(executor): Fix compatibility issue when selfLink is no longer populated for k8s>=1.21. Fixes #6045

### LINKED ISSUE #6045: workflow-of-workflows : waiting "child" workflow is not working
## Summary 

I try to use https://github.com/argoproj/argo-workflows/blob/master/examples/workflow-of-workflows.yaml

The child workflow is executed correctly. However, the parent fails to wait with a strange error (see the logs)

From my understanding, the problem is due to `Failing for conditions: status.phase in (Error,Failed)` with `index out of range [-1]` which is coming from the initial yaml as `failureCondition: "status.phase in (Failed, Error)"` 

However, this is directly from the provided example ->is this something wrong on my side or is this a bug ?

Here are the logs
```
time="2021-05-28T14:26:54.792Z" level=info msg="Starting Workflow Executor" executorType=k8sapi version=untagged
time="2021-05-28T14:26:54.796Z" level=info msg="Creating a K8sAPI executor"
time="2021-05-28T14:26:54.796Z" level=info msg="Executor initialized" includeScriptOutput=false namespace=staging-argo podName=dataset-all-in-coverage-2crft-3484945999 template="{\"name\":\"dataset-all-in-coverage-execute\",\"inputs\":{\"paramete
rs\":[{\"name\":\"coverage\",\"value\":\"fr-cor\"},{\"name\":\"dataset\",\"value\":\"ajaccio\"}]},\"outputs\":{},\"metadata\":{},\"resource\":{\"action\":\"create\",\"manifest\":\"\\napiVersion: argoproj.io/v1alpha1\\nkind: Workflow\\nmetadata:\\
n  generateName: dataset-\\nspec:\\n  workflowTemplateRef:\\n    name: dataset\\n  arguments:\\n    parameters:\\n    - name: dataset\\n      value: ajaccio\\n    - name: coverage\\n      value: fr-cor\\n\",\"successCondition\":\"status.phase ==
Succeeded\",\"failureCondition\":\"status.phase in (Failed, Error)\"},\"archiveLocation\":{\"archiveLogs\":true,\"s3\":{\"endpoint\":\"minio-az-gw:9000\",\"bucket\":\"argo\",\"insecure\":true,\"accessKeySecret\":{\"name\":\"argo-storage-secret\",
\"key\":\"accessKey\"},\"secretKeySecret\":{\"name\":\"argo-storage-secret\",\"key\":\"secretKey\"},\"key\":\"dataset-all-in-coverage-2crft/dataset-all-in-coverage-2crft-3484945999\"}}}" version="&Version{Version:untagged,BuildDate:2021-05-27T23:
52:25Z,GitCommit:73539fadbe81b644b912ef0ddddebb178c97cc94,GitTag:untagged,GitTreeState:clean,GoVersion:go1.15.7,Compiler:gc,Platform:linux/amd64,}"
time="2021-05-28T14:26:54.796Z" level=info msg="Loading manifest to /tmp/manifest.yaml"
time="2021-05-28T14:26:54.796Z" level=info msg="kubectl create -f /tmp/manifest.yaml -o json"
time="2021-05-28T14:26:55.794Z" level=info msg="Resource: staging-argo/Workflow.argoproj.io/dataset-2zlqn. SelfLink: "
time="2021-05-28T14:26:55.794Z" level=info msg="Starting SIGUSR2 signal monitor"
time="2021-05-28T14:26:55.795Z" level=info msg="Waiting for conditions: status.phase==Succeeded"
time="2021-05-28T14:26:55.795Z" level=info msg="Failing for conditions: status.phase in (Error,Failed)"
E0528 14:26:55.818725       1 runtime.go:78] Observed a panic: runtime.boundsError{x:-1, y:1, signed:true, code:0x0} (runtime error: index out of range [-1])
goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2022ae0, 0xc000436b20)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.19.6/pkg/util/runtime/runtime.go:48 +0x89
panic(0x2022ae0, 0xc000436b20)
        /usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/argoproj/argo-workflows/v3/util/k8s.ParseRequest(0xc00070e400, 0xc00070e400, 0xc00084c090, 0x0, 0x0)
        /go/src/github.com/argoproj/argo-workflows/util/k8s/parse.go:45 +0x739
github.com/argoproj/argo-workflows/v3/util/logs.k8sLogRoundTripper.RoundTrip(0x245c400, 0xc00016b040, 0xc00070e400, 0x1e15c40, 0xc000714a01, 0xc00070c420)
        /go/src/github.com/argoproj/argo-workflows/util/logs/log-k8s-requests.go:19 +0xb2
k8s.io/client-go/transport.(*bearerAuthRoundTripper).RoundTrip(0xc00033c690, 0xc00070e300, 0x216157e, 0xa, 0xc00070c360)
        /go/pkg/mod/k8s.io/client-go@v0.19.6/transport/round_trippers.go:297 +0x271
k8s.io/client-go/transport.(*userAgentRoundTripper).RoundTrip(0xc0001a3760, 0xc00070e200, 0xc0001a3760, 0xc0245d53ef659ab4, 0x7b1191738)
        /go/pkg/mod/k8s.io/client-go@v0.19.6/transport/round_trippers.go:159 +0x1d6
net/http.send(0xc00070e100, 0x245c1a0, 0xc0001a3760, 0xc0245d53ef659ab4, 0x7b1191738, 0x3235b80, 0xc000182020, 0xc0245d53ef659ab4, 0x1, 0x0)
        /usr/local/go/src/net/http/client.go:252 +0x453
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by terrytangyuan**
See context in https://github.com/argoproj/argo-workflows/issues/6008. Also Fixes #6045. 

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

**Comment by terrytangyuan** [comment]
@alexec The new test failed for only some of the executors. Is there any difference among the executors in CI environment that I may be missing?

**Comment by terrytangyuan** [comment]
I am seeing the following in controller log: 

logs | k8s-resource-tmpl-with-wf-m6jjx main Error: signal: terminated
1050
 controller | time="2021-05-26T04:37:38.961Z" level=info msg="Create pods/exec 500"
1051
 controller | time="2021-05-26T04:37:38.962Z" level=warning msg="failed to clean-up pod" action=terminateContainers error= key=argo/k8s-resource-tmpl-with-wf-m6jjx/terminateContainers

@alexec Only tests for a few of the executors failed. Is there anything special about the CI environment for different executors that  I should be aware of?

**Comment by alexec** [comment]
Step 1. Let's re-run the jobs.

**Comment by terrytangyuan** [comment]
Okay tried a couple of things but those are not very helpful. It looks like each time different executors failed so there maybe something unstable hidden in the CI environment itself.

**Comment by terrytangyuan** [comment]
I've located the root cause which will be fixed in https://github.com/argoproj/argo-workflows/pull/6033. I'll clean up this PR once #6033 is merged so this PR can focus on supporting k8s>=1.21.

**Comment by terrytangyuan** [comment]
This PR has been rebased after https://github.com/argoproj/argo-workflows/pull/6033 got merged. @alexec Feel free to review whenever you get a chance.

File: test/e2e/resource_template_test.go [hunk 1]
```diff
@@ -54,6 +54,45 @@ spec:
...
+        manifest: |
+          apiVersion: v1
+          kind: Pod
+          metadata:
+            generateName: k8s-pod-resource-
+          spec:
+            serviceAccountName: argo
+            containers:
+            - name: pi
+              image: perl
```
Comment (alexec) [inline test/e2e/resource_template_test.go:83]: should not use a special image, makes builds slow, can you use a whitelisted image please, e.g `argoproj/argosay:v2`

Comment (terrytangyuan) [reply test/e2e/resource_template_test.go]: Sounds good. Done.

Comment (terrytangyuan) [reply test/e2e/resource_template_test.go]: Sounds good. Changed.

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
Lets hold this for v3.2

**Comment by christopheblin** [comment]
@alexec 

> Lets hold this for v3.2

Without this fix, "workflow-of-workflows" is not working with k8sapi executor under "new" k8s versions (basically, I think any resource creation that takes a bit of time is not working, so it means Workflows but also Deployments)

Can you explain why the fix is hold ? Will it be available in latest or in a 3.1.1 soon ?

Or is there a workaround to "wait" the resource in a next step ?

**Comment by alexec** [comment]
Are we saying that self link issue is already impacting production systems? I.e. 1.21 is now running in production?

**Comment by christopheblin** [comment]
Thanks  for the merge @alexec  👍 

I am running Azure Kubernetes Services (AKS) which, by the analysis of logs by @terrytangyuan in #6045, has the selfLink issue

I'll try to test the "latest" image tomorrow (CET) and I'll come back with a new issue if needed

**Pull request merged**